### DISCUSSION TITLE
Identation defining the scope

### DISCUSSION THREAD

**Discussion by JoanGi** [eclipse-langium/langium#782] [Q&A] [answered]
Hi,

In my use case will be a great feature that indentation defining the scope of autocompletion, validation etc... I was not able to guess if this is possible or how can it be done. 

For example: If I'm in POSITION 1, I can set the “Type:"  keyword, but If I'm in POSITION 2, I only can set the "Dependencies:" keyword of the parent class.

Anyone has any clue to achieve this? Thanks in advance! :)

```
// Grammar
 Composition:
     name='Composition:'
     'Rationale:' compodesc+=STRING
     ('Total size:' numberInst=NUMBER)? 
     instances = DataInstance
     ('Dependencies:' dependencies=Dependencies)? 

DataInstance:
    'Instance:' name=ID 
    'Description:' descript=STRING 
    'Type:'  type=InstanceType  

```
Language instace

```
Composition:
     Rationale: ""
     Total size: 1143
     Data Instances:
         Instance:  Names
             Description: "Describe the instance"
             [POSITION 1]
         [POSITION 2]
```

**Comment by msujew** [comment, marked as answer]
Hey @JoanGi,

So, this can be done in theory, by using whitespace sensitive approach to parsing. You would introduce "decoy" terminal rules `INDENT` and `DEDENT`, then you override their `pattern` in the [`TokenBuilder`](https://github.com/langium/langium/blob/a18ebca4ce58391d1d7bbb7dc1509b6ec597bb23/packages/langium/src/parser/token-builder.ts#L46-L53) class to look something like [this chevrotain python example](https://github.com/chevrotain/chevrotain/blob/master/examples/lexer/python_indentation/python_indentation.js). You would then use these tokens in your grammar whenever you need to increase/decrease the indentation of your text. This approach works best, when you enter a new parser rule/create a new AST node for each level of indentation, as that later on massively helps you with validation and cross reference resolution.

Note that whitespace sensitive languages in EBNF style grammars are always a pain to deal with, and there's no easy for accomplishing this.

**Reply by aabounegm** [reply]
Update: the [IndentationAwareTokenBuilder](https://eclipse-langium.github.io/langium/classes/langium.IndentationAwareTokenBuilder.html) has been released in v3.2.0, and there is now [a recipe](https://langium.org/docs/recipes/lexing/indentation-sensitive-languages/) explaining how to use it.
